### ISSUE TITLE
Out of memory on commitTransaction

### ISSUE THREAD

**Issue by mandrachek** [realm/realm-java#3050]
Here's the stack trace I'm getting:

> Fatal Exception: io.realm.exceptions.RealmError: Unrecoverable error. mmap() failed: Out of memory size: 402653184 in io_realm_internal_SharedGroup.cpp line 188
>        at io.realm.internal.SharedGroup.nativeCommitAndContinueAsRead(SharedGroup.java)
>        at io.realm.internal.SharedGroup.commitAndContinueAsRead(SharedGroup.java:95)
>        at io.realm.internal.ImplicitTransaction.commitAndContinueAsRead(ImplicitTransaction.java:62)
>        at io.realm.internal.SharedGroupManager.commitAndContinueAsRead(SharedGroupManager.java:118)
>        at io.realm.BaseRealm.commitTransaction(BaseRealm.java:340)
>        at io.realm.BaseRealm.commitTransaction(BaseRealm.java:327)
>        at io.realm.Realm.commitTransaction(Realm.java:120)
>        at myapp.services.ToDoIntentService.getTasks(ToDoIntentService.java:350)
>        at myapp.services.ToDoIntentService.doWork(ToDoIntentService.java:70)
>        at myapp.services.RealmBackgroundIntentService.onHandleIntent(RealmBackgroundIntentService.java:56)
>        at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:65)
>        at android.os.Handler.dispatchMessage(Handler.java:102)
>        at android.os.Looper.loop(Looper.java:135)
>        at android.os.HandlerThread.run(HandlerThread.java:61)

I have a class that I use to monitor what classes have opened a realm, and if they've closed it or not. I use this exclusively over Realm.getDefaultInstance()/close().  I've had lots of problems with fragments not shutting down in a timely fashion and so not closing their realm instances.

```
public class RealmMonitor {

    private static volatile RealmMonitor _instance;
    private final Map<String,Integer> map = Collections.synchronizedMap(new HashMap<String, Integer>());
    private final Object mapLock = new Object();

    private RealmMonitor() {

    }

    protected static RealmMonitor getInstance() {
        if (_instance == null) {
            synchronized (RealmMonitor.class) {
                if (_instance == null) {
                    _instance = new RealmMonitor();
                }
            }
        }
        return _instance;
    }

    public static Realm getDefaultInstance(Object o) throws RealmMigrationNeededException, RealmError, RealmIOException {
        //Timber.e("REALM acquired: " + o.getClass().getName());
        Realm realm = Realm.getDefaultInstance();
        incrementCount(o);
        return realm;
    }

    public static void close(Object o, Realm realm) {
        if (realm != null && !realm.isClosed()) {
            //Timber.e("REALM released: " + o.getClass().getName());
            try {
                Looper looper = Looper.myLooper();
                if (looper != null) {
                    realm.removeAllChangeListeners();
                }
            }
            catch (Exception e) {
                Timber.e(e, "Realm change listeners could not be removed.");
            }
            try {
                RealmConfiguration configuration = realm.getConfiguration();
                realm.close();
                decrementCount(o);
            }
            catch (Exception e) {
                Timber.e(e, "Realm could not be closed.");
            }
        }
    }


    private static void incrementCount(Object o) {
        String caller = o.getClass().getName();
        RealmMonitor monitor = getInstance();
        synchronized (monitor.mapLock) {
            if (!monitor.map.containsKey(caller)) {
                monitor.map.put(caller, 1);
            } else {
                monitor.map.put(caller, monitor.map.get(caller) + 1);
            }
        }
    }

    private static void decrementCount(Object o) {
        String caller = o.getClass().getName();
        RealmMonitor monitor = getInstance();
        synchronized (monitor.mapLock) {
            if (monitor.map.containsKey(caller)) {
                monitor.map.put(caller, monitor.map.get(caller) -1);
            }
        }
    }

    private static String getCaller() {
        StackTraceElement[] traceElements = Thread.currentThread().getStackTrace();
        String caller = null;
        for (StackTraceElement traceElement : traceElements) {
            Timber.e(traceElement.getClassName());
        }
        for (StackTraceElement traceElement : traceElements) {
            caller = traceElement.getClassName();
            if (
                    !caller.equals(RealmMonitor.class.getName()) &&
                            !caller.equals(Thread.class.getName()) &&
                            !caller.startsWith("android") &&
                            !caller.startsWith("dalvik")
                    ) {
                break;
            }
        }
        return caller;
    }

    public static void log() {
        if (BuildConfig.DEBUG) {
            synchronized (getInstance().mapLock) {
                Timber.d("========================== REALM MONITOR =======================");
                for (Map.Entry<String, Integer> entry : getInstance().map.entrySet()) {
                    Timber.d("| " + entry.getKey() + "\t - " + String.valueOf(entry.getValue()) + " Connection(s)");
                }
                Timber.d("========================== REALM MONITOR =======================");
            }
        }
    }

}
```

I have a RealmBackgroundIntentService that looks like this:
```
public abstract class RealmBackgroundIntentService extends IntentService {

    protected volatile Realm realm = null;
    protected final Object realmLock = new Object();


    public RealmBackgroundIntentService(String name) {
        super(name);
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        PowerManager powerManager = (PowerManager)getSystemService(POWER_SERVICE);
        PowerManager.WakeLock wl = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK,"lock");
        wl.acquire();
        try {
            android.os.Process.setThreadPriority(Process.THREAD_PRIORITY_BACKGROUND);

            if (realm == null) {
                synchronized (realmLock) {
                    if (realm == null) {
                        try {
                            realm = RealmMonitor.getDefaultInstance(this);
                        }
                        catch (RealmMigrationNeededException rmne) {
                            Timber.e(rmne, "Realm Migration Needed.");
                            Toast.makeText(getApplicationContext(), "You need to re-install. Thanks.", Toast.LENGTH_LONG).show();
                            return;
                        }
                        catch (Exception e) {
                            Timber.e(e,"Couldn't open realm.");
                        }
                    }
                }
            }

            doWork(intent);

        }
        catch (Exception e) {
            Timber.e(e, "An error occurred attempting to perform intent.");
        }
        finally {
            synchronized (realmLock) {
                if (realm != null) {
                    RealmMonitor.close(this, realm);
                    realm = null;
                }
            }
            try {
                if (wl.isHeld()) {
                    wl.release();
                }
            }
            catch (Exception e) {
                Timber.e(e,"Couldn't release wakelock");
            }
        }
    }



    public abstract void doWork(Intent intent);
}
```

The service mentioned in the stack trace, looks like this:
```
public class ToDoIntentService extends RealmBackgroundIntentService {
    private static final String TAG = ToDoIntentService.class.getSimpleName();

    @Inject
    RetrofitService retrofitService;

    public ToDoIntentService() {
        super(TAG);
    }

    @Override
    public void onCreate() {
        super.onCreate();
        MyApplication.getComponent(this).inject(this);
    }

    // implements the doWork from RealmBackgroundIntentService
    @Override
    public void doWork(Intent intent) {
            getTasks(intent);
    }


    private void getTasks(Intent intent) {
        final String id = intent.getStringExtra(Actions._ID);

        EventBus.getDefault().postSticky(new CVEvents.Tasks(CVEvents.STARTED));

        try {
            final Call<List<Task>> getTasksCall = retrofitService.getTasks(villageId);
            final Response<List<Task>> getTasksResponse = getTasksCall.execute();

            if (getTasksResponse.isSuccessful()) {
                final List<Task> tasks = getTasksResponse.body();
                realm.beginTransaction();
                realm.copyToRealmOrUpdate(tasks);
                realm.commitTransaction();
                EventBus.getDefault().postSticky(new CVEvents.Tasks(CVEvents.SUCCESS));
            }
            else {
                EventBus.getDefault().postSticky(new CVEvents.Tasks(logException(getTasksResponse)));
            }
        }
        catch (Exception e) {
            EventBus.getDefault().postSticky(new CVEvents.Tasks(logException(e)));
        }

    }
}
```

This IntentService gets called 1 of 3 ways:
1) A push message comes in that says I need to update
2) When a user pulls to refresh on the tasks screen.
3) During the initial data loading process. 

For #3, I have an intent service that makes 4-5 http calls, and inserts data into the realm.
Once those 4-5 initial calls are done, there's about 8 "child" intent services that get called, per "item" from one of those initial calls (generally people have 1-5 items). Something like this:

```
for (Item item : items) {
   String id = item.getId();
   Actions.getMembers(this,id);
   Actions.getTasks(this,id);
   Actions.getEvents(this,id);
   Actions.getAnswers(this,id);
   Actions.getWeather(this,id);
   Actions.getJournals(this,id);
}
```

Unfortunately with the way IntentServices work, at least with this crash, I don't have any way to determine where it's being called.

For most users the app's data sits at 1-2MB.  I'm checking with the user that experienced this crash to find out what their Data size is.

In everything I've seen with my RealmMonitor, all my RealmBackgroundIntentServices open and close their realms as expected.  All my Activities open and close their realms as expected.  All my fragments open their realms as expected, and eventually close them, as expected.

What could be going on here?

#### Version of Realm and tooling
Realm version(s): 1.0.0
Android Studio version: 2.1
Which Android version and device: 5.1.1, Nexus 7 (2nd gen)

**Comment by Zhuinden** [comment]
My initial guesses are:

1.) at least one of your Realm stays open on the background thread
2.) there are too many simultaneous transactions open on background threads
3.) your locking prevents Realms from being closed?

I'm not entirely sure why you need the `RealmMonitor`, Realm already ensures that you can access it only on the thread where you've opened it, and is otherwise threadsafe.

**Comment by mandrachek** [comment]
1) I don't believe this is the case. The realm monitor pretty clearly shows this. All my services and activities close their realms properly. My fragments close their realms properly too (eventually. :))

2) Given that IntentServices operate as a queue (only one instance of an intent service's onHandleIntent is supposed to run at a time), and the transactions are also limited by waiting for HTTP requests (and getting an HTTP connection from the connection pool) the max concurrent transactions services would be 8. But it's probably less than that in the vast majority of cases.

3) No, the locking doesn't prevent the realm from being closed. I'm not getting any lock wait timeout exceptions, and all the IntentService's clear out from the RealmMonitor pretty quickly. The only way there should ever be contention for the lock in the IntentService is if the onHandleIntent on a particular service instance was being called twice (which _shouldn't_ ever happen).  Is that really too many?

I've also tried compacting the realm, but my realm is encrypted, and apparently that's not an option.

FWIW, the RealmMonitor was created because I wanted to delete the realm and start from scratch when a user logs out. The only issue is, I couldn't ever seem to get all the realm instances to close so that I could do this.  So I added the RealmMonitor to keep track of what classes had opened a realm, and how many instances were open, so I could see where the realm instances were being held. Eventually the fragments do give up their instances, but the amount of time it takes varies greatly. I eventually gave up on deleting the realm, and just clear the data (which isn't ideal in my case, but works).

**Comment by leisim** [comment]
I have the exact same problem :/

**Comment by beeender** [comment]
Do you ever get "Realm could not be closed." in your adb logcat?

**Comment by mandrachek** [comment]
No, I've never seen that show up.

**Comment by mandrachek** [comment]
Unfortunately the user that encountered this error uninstalled the app and installed an update before sending me a screenshot of the data size. :(  I haven't been able to reproduce this, it's only happened as far as I know, to a single user, a single time at this point.

**Comment by leisim** [comment]
My database was 800mb but has less than 50 entries o_O

**Comment by Zhuinden** [comment]
@leisim  you're also not closing the Realm on your background threads.

**Comment by kneth** [comment]
@leisim As @Zhuinden points out, closing Realms is important: https://realm.io/docs/java/latest/#controlling-the-lifecycle-of-realm-instances

**Comment by leisim** [comment]
I triple-checked every realm instance :/

**Comment by kneth** [comment]
@leisim Can you create a small project which produce the problem? Or can you share your project either in public or private (just send an email to help@realm.io)?

**Comment by leisim** [comment]
@kneth The problem is that I can't even reproduce the error myself. Whatever I try, the realm file keeps its size and does not grow. And on multiple devices running my app at some time it is just over 500mb big. I'm kind of desperate.

**Comment by Zhuinden** [comment]
@leisim do you by chance have multiple intent services writing to the Realm at the same time, which are in parallel and get blocked if there is an issue with internet connectivity for example?

I also recommend calling `compactRealm()` when all of your Activities are closed, although it should not grow to 800 MBs unless background thread versions are retained.

**Comment by kneth** [comment]
@leisim Having multiple threads blocked is the root of the problem. Realm is a multi-version database, so when you commit in one thread and thereby creates a new version, the other threads are still holding on the previous versions. We know that many apps are created this way, and that's why we have `compactRealm()` :-)

**Comment by leisim** [comment]
@kneth Is it possible to disable the multi version behaviour? My app runs many weeks on a kiosk station without being restarted and it is hard to call `compactRealm()` in this time...

**Comment by Zhuinden** [comment]
I think the workaround would be to use looper threads as blocking worker threads, considering that the older versions are not retained if the Realm autoupdates on the threads if I know right.

**Comment by leisim** [comment]
@Zhuinden Is there a method like `compactRealm()` which I can call when the thread finished its work to clean the old versions for this thread?

**Comment by Zhuinden** [comment]
... Technically that is `compactRealm()` which works only if NO thread has the Realm open. 

I think converting the background threads into looper threads might be the easiest way to minimize the number of retained versions in your case, but I am not actually a Realm person and this is my hypothesis, I am not actually sure. 

@cmelchior are old versions still retained if the "background threads" are capable of autoupdating using their own dedicated Looper?

**Comment by kneth** [comment]
@leisim That's an interesting use case. Of course, blocking code should be in background threads. `compactRealm()` sounds a bit hard in your use case. Since `compactRealm()` updates the internal data structures in our storage engine, it is a requirement that only one instance is open.

By using a Lopper thread, your can get your instances updates and thereby freeing/reusing allocated memory.

**Comment by leisim** [comment]
@kneth Converting all background threads to Looper threads is not a good solution in my optinion. Is there anything else I can do?

**Comment by Zhuinden** [comment]
@leisim well there's the `waitForChange()` method which blocks a thread and updates it.

Previously it used to be `refresh()` which instantly brought the thread up to speed, and does that using `sharedGroupManager.advanceRead()` in the Realm code.

**Comment by cmelchior** [comment]
@leisim If you are seeing the file size explode, most likely you are not closing the Realm somewhere. It would be really helpful if you could share your code. It can be done privately through help@realm.io if you wish.

**Comment by leisim** [comment]
I added @cmelchior and @kneth as collaborators to the github repo...
The app which has these problems is the 'koessler' module. And the 'common' module is our library.
If you have any questions please feel free to ask.

**Comment by kneth** [comment]
@leisim Thanks

**Comment by kneth** [comment]
I will close the issue. I believe @leisim got a few advice by direct email, and @mandrachek has only seen it once. If you see it again, please either reopen or create new issue.
